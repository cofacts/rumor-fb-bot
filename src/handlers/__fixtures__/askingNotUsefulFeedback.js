export const oneFeedback = {
  data: {
    action: {
      feedbackCount: 1,
    },
  },
};

export const twoFeedbacks = {
  data: {
    action: {
      feedbackCount: 2,
    },
  },
};
