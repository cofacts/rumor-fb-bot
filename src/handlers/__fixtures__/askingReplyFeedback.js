export const twoFeedbacks = {
  data: {
    action: {
      feedbackCount: 2,
    },
    GetReply: {
      type: 'NOT_ARTICLE',
      text:
        '這是商業活動廣告01 這是商業活動廣告02 這是商業活動廣告03 這是商業活動廣告04 這是商業活動廣告05 這是商業活動廣告06 這是商業活動廣告07 這是商業活動廣告08 這是商業活動廣告09 這是商業活動廣告10 這是商業活動廣告11 這是商業活動廣告12 這是商業活動廣告13 這是商業活動廣告14 這是商業活動廣告15 這是商業活動廣告16 這是商業活動廣告17 這是商業活動廣告18 這是商業活動廣告19 這是商業活動廣告20',
      reference: '',
    },
    GetArticle: {
      text:
        '(0)(1)(/)(0)(9)(line)免費貼圖\n\n「[全螢幕貼圖]生活市集x生活小黑熊」\nhttps://line.me/S/sticker/10299\n\n「松果購物 x 狗與鹿(動態貼圖) 」\nhttps://line.me/S/sticker/10300\n\n「LINE TV X 我的男孩限免貼圖」\nhttps://line.me/S/sticker/10207',
    },
  },
};

export const oneFeedback = {
  data: {
    action: {
      feedbackCount: 1,
    },
  },
};

export const articleData = {
  data: {
    GetReply: {
      type: 'NOT_ARTICLE',
      text:
        '這是商業活動廣告01 這是商業活動廣告02 這是商業活動廣告03 這是商業活動廣告04 這是商業活動廣告05 這是商業活動廣告06 這是商業活動廣告07 這是商業活動廣告08 這是商業活動廣告09 這是商業活動廣告10 這是商業活動廣告11 這是商業活動廣告12 這是商業活動廣告13 這是商業活動廣告14 這是商業活動廣告15 這是商業活動廣告16 這是商業活動廣告17 這是商業活動廣告18 這是商業活動廣告19 這是商業活動廣告20',
      reference: '',
    },
    GetArticle: {
      text:
        '(0)(1)(/)(0)(9)(line)免費貼圖\n\n「[全螢幕貼圖]生活市集x生活小黑熊」\nhttps://line.me/S/sticker/10299\n\n「松果購物 x 狗與鹿(動態貼圖) 」\nhttps://line.me/S/sticker/10300\n\n「LINE TV X 我的男孩限免貼圖」\nhttps://line.me/S/sticker/10207',
    },
  },
};
