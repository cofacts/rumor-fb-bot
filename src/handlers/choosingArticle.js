import { t } from 'ttag';
import gql from '../gql';
import {
  createPostbackAction,
  createFeedbackWords,
  createTypeWords,
  isNonsenseText,
  getArticleURL,
  createAskArticleSubmissionReply,
} from './utils';
import ga from '../ga';

/**
 * We reorder the replies queried and put them into a reply list.
 * Replies indicating truth or rumor are listed first.
 * If the reply list is not yet full, we then add the 'NOT_ARTICLE' replies.
 * If this article has more than 10 replies, we put a 'read more' block
 * that links to our website at the end of the reply list.
 */
function reorderArticleReplies(articleReplies) {
  const replies = [];
  const notArticleReplies = [];

  for (let articleReply of articleReplies) {
    if (articleReply.reply.type !== 'NOT_ARTICLE') {
      replies.push(articleReply);
    } else {
      notArticleReplies.push(articleReply);
    }
  }
  return replies.concat(notArticleReplies);
}

/**
 * The state that we're processing user input,
 * and the the user should choose which article matches what she's seeking.
 * If she hasn't chosen one, we ask her to select the desired reply.
 * If she has, we query replies of the article and ask her to choose those she wants to read.
 * If we didn't find any reply, we ask the user to tell us why she thinks this should be replied sooner.
 */
export default async function choosingArticle(params) {
  let { data, state, event, issuedAt, userId, replies, isSkipUser } = params;

  if (!data.foundArticleIds) {
    throw new Error('foundArticleIds not set in data');
  }

  data.selectedArticleId = data.foundArticleIds[event.input - 1];
  const { selectedArticleId } = data;
  const doesNotContainMyArticle = +event.input === 0;

  if (doesNotContainMyArticle && isNonsenseText(data.searchedText)) {
    replies = [
      {
        type: 'text',
        content: {
          text:
            t`Sorry, please provide more information.` +
            '\n' +
            t`Please refer to our user's manual` +
            ' üìñ http://bit.ly/cofacts-fb-users',
        },
      },
    ];
    state = '__INIT__';
  } else if (doesNotContainMyArticle) {
    replies = createAskArticleSubmissionReply();

    state = 'ASKING_ARTICLE_SUBMISSION_REASON';
  } else if (!selectedArticleId) {
    replies = [
      {
        type: 'text',
        content: {
          text: t`Please enter 1ÔΩû${
            data.foundArticleIds.length
          } to choose a reply.`,
        },
      },
    ];

    state = 'CHOOSING_ARTICLE';
  } else {
    const {
      data: { GetArticle },
    } = await gql`
      query($id: String!) {
        GetArticle(id: $id) {
          replyCount
          text
          articleReplies(status: NORMAL) {
            reply {
              id
              type
              text
            }
            positiveFeedbackCount
            negativeFeedbackCount
          }
        }
      }
    `({
      id: selectedArticleId,
    });

    data.selectedArticleText = GetArticle.text;
    const visitor = ga('FB-' + userId, data.selectedArticleText);
    visitor.screenview({ screenName: state });

    // Track which Article is selected by user.
    visitor.event({
      ec: 'Article',
      ea: 'Selected',
      el: selectedArticleId,
      dt: data.selectedArticleText,
    });

    const count = {
      RUMOR: 0,
      NOT_RUMOR: 0,
      OPINIONATED: 0,
      NOT_ARTICLE: 0,
    };

    GetArticle.articleReplies.forEach(ar => {
      // Track which Reply is searched. And set tracking event as non-interactionHit.
      visitor.event({ ec: 'Reply', ea: 'Search', el: ar.reply.id, ni: true });

      const type = ar.reply.type;
      count[type] += 1;
    });

    const articleReplies = reorderArticleReplies(GetArticle.articleReplies);
    const summary =
      t`${count.RUMOR} person(s) consider this to be a rumor` +
      ' ‚ùå\n' +
      t`${count.NOT_RUMOR} person(s) think this can be a truth` +
      ' ‚≠ï\n' +
      t`${count.OPINIONATED} person(s) think this is simply personal opinion` +
      ' üí¨\n' +
      t`${
        count.NOT_ARTICLE
      } person(s) thinks it's off-topic and Cofacts need not to handle this message` +
      ' ‚ö†Ô∏èÔ∏è\n';

    replies = [
      {
        type: 'text',
        content: {
          text: summary,
        },
      },
    ];

    if (articleReplies.length !== 0) {
      data.foundReplyIds = articleReplies.map(({ reply }) => reply.id);

      state = 'CHOOSING_REPLY';

      if (articleReplies.length === 1) {
        // choose for user
        event.input = 1;

        return {
          data,
          state: 'CHOOSING_REPLY',
          event,
          issuedAt,
          userId,
          replies,
          isSkipUser: true,
        };
      }

      const templateMessage = {
        type: 'template',
        //altText: createAltText(articleReplies),
        payload: {
          template_type: 'generic',
          elements: articleReplies
            .slice(0, 10)
            .map(
              (
                { reply, positiveFeedbackCount, negativeFeedbackCount },
                idx
              ) => ({
                subtitle:
                  createTypeWords(reply.type) +
                  '\n' +
                  createFeedbackWords(
                    positiveFeedbackCount,
                    negativeFeedbackCount
                  ),
                title: reply.text.slice(0, 80),
                buttons: [createPostbackAction(t`Read this reply`, idx + 1)],
              })
            ),
        },
      };
      replies.push({
        type: 'carousel',
        content: {
          attachment: templateMessage,
        },
      });

      const articleUrl = getArticleURL(selectedArticleId);
      if (articleReplies.length > 10) {
        replies.push({
          type: 'text',
          content: {
            text: t`Please refer to ${articleUrl} for more replies.`,
          },
        });
      }
    } else {
      // No one has replied to this yet.

      // Track not yet reply Articles.
      visitor.event({
        ec: 'Article',
        ea: 'NoReply',
        el: selectedArticleId,
      });

      const replyText =
        '„Äê' +
        t`Tell us about your concern` +
        '„Äë\n' +
        t`Sorry, no one has replied to this article yet!` +
        '\n\n' +
        t`If you consider this a rumor, please tell us your concern and why we should figure this out as soon as possible.` +
        '\n\n';
      const promptText =
        t`Please send us in messages the reason why you consider this a rumor.` +
        '\n';

      replies = [
        {
          type: 'text',
          content: {
            text: replyText,
          },
        },
        {
          type: 'text',
          content: {
            text: promptText,
          },
        },
        {
          type: 'buttons',
          content: {
            attachment: {
              type: 'template',
              payload: {
                template_type: 'button',
                text: t`To skip this, click "Skip"`,
                buttons: [createPostbackAction(t`Skip`, 'n')],
              },
            },
          },
        },
      ];

      state = 'ASKING_REPLY_REQUEST_REASON';
    }
    visitor.send();
  }

  return { data, state, event, issuedAt, userId, replies, isSkipUser };
}
