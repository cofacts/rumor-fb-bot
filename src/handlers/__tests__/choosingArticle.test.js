jest.mock('../../gql');

import choosingArticle from '../choosingArticle';
import * as apiResult from '../__fixtures__/choosingArticle';
import gql from '../../gql';

it('should select article by articleId', async () => {
  gql.__push(apiResult.selectedArticleId);

  const params = {
    data: {
      searchedText:
        '《緊急通知》\n台北馬偕醫院傳來訊息：\n資深醫生（林清風）傳來：「請大家以後千萬不要再吃生魚片了！」\n因為最近已經發現- 好多病人因為吃了生魚片，胃壁附著《海獸胃腺蟲》，大小隻不一定，有的病人甚至胃壁上滿滿都是無法夾出來，驅蟲藥也很難根治，罹患機率每個國家的人都一樣。\n尤其；鮭魚的含蟲量最高、最可怕！\n請傳給朋友，讓他們有所警惕!',
      foundArticleIds: [
        'AVyyB61NyCdS-nWhuakC',
        'AV4sEcSHyCdS-nWhufEX',
        '5478658696099-rumor',
      ],
    },
    state: 'CHOOSING_ARTICLE',
    event: {
      type: 'message',
      input: '1',
      timestamp: 1505314293406,
      message: {
        type: 'text',
        id: '6691804381697',
        text: '1',
      },
    },
    issuedAt: 1505314295017,
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    isSkipUser: false,
  };

  expect(await choosingArticle(params)).toMatchSnapshot();
});

it('should select article and have OPINIONATED and NOT_ARTICLE replies', async () => {
  gql.__push(apiResult.multipleReplies);

  const params = {
    data: {
      searchedText:
        '老榮民九成存款全部捐給慈濟，如今窮了卻得不到慈濟醫院社工的幫忙，竟翻臉不認人',
      foundArticleIds: ['AV8d2-YtyCdS-nWhuhdi'],
    },
    state: 'CHOOSING_ARTICLE',
    event: {
      type: 'message',
      input: '1',
      timestamp: 1511633232479,
      message: { type: 'text', id: '7045918737413', text: '1' },
    },
    issuedAt: 1511633232970,
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  expect(await choosingArticle(params)).toMatchSnapshot();
});

it('should select article with no replies', async () => {
  gql.__push(apiResult.noReplies);
  gql.__push(apiResult.createReplyRequest);

  const params = {
    data: {
      searchedText: '老司機車裡總備一塊香皂，知道內情的新手默默也準備了一塊',
      foundArticleIds: ['AV_4WX8vyCdS-nWhujyH'],
    },
    state: 'CHOOSING_ARTICLE',
    event: {
      type: 'message',
      input: '1',
      timestamp: 1511702208226,
      message: { type: 'text', id: '7049700770815', text: '1' },
    },
    issuedAt: 1511702208730,
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  expect(await choosingArticle(params)).toMatchSnapshot();
});

it('should select article and slice replies when over 10', async () => {
  gql.__push(apiResult.elevenReplies);

  const params = {
    data: {
      searchedText:
        '老榮民九成存款全部捐給慈濟，如今窮了卻得不到慈濟醫院社工的幫忙，竟翻臉不認人',
      foundArticleIds: ['AV8d2-YtyCdS-nWhuhdi'],
    },
    state: 'CHOOSING_ARTICLE',
    event: {
      type: 'message',
      input: '1',
      timestamp: 1511633232479,
      message: { type: 'text', id: '7045918737413', text: '1' },
    },
    issuedAt: 1511633232970,
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  expect(await choosingArticle(params)).toMatchSnapshot();
});

it('should ask users if they want to submit article when user say not found', async () => {
  const params = {
    data: {
      searchedText:
        '這一篇文章確實是一個轉傳文章，他夠長，看起來很轉傳，但是使用者覺得資料庫裡沒有。',
      foundArticleIds: ['AV8d2-YtyCdS-nWhuhdi'],
    },
    state: 'CHOOSING_ARTICLE',
    event: {
      type: 'message',
      input: '0',
      timestamp: 1511633232479,
      message: { type: 'text', id: '7045918737413', text: '0' },
    },
    issuedAt: 1511633232970,
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  const result = await choosingArticle(params);
  expect(result).toMatchSnapshot();
});

it('should block user from submitting articles that is too short', async () => {
  const params = {
    data: {
      searchedText: '這真的假的？？！！！？？！',
      foundArticleIds: ['AV8d2-YtyCdS-nWhuhdi'],
    },
    state: 'CHOOSING_ARTICLE',
    event: {
      type: 'message',
      input: '0',
      timestamp: 1511633232479,
      message: { type: 'text', id: '7045918737413', text: '0' },
    },
    issuedAt: 1511633232970,
    userId: 'Uc76d8ae9ccd1ada4f06c4e1515d46466',
    replies: undefined,
    isSkipUser: false,
  };

  const result = await choosingArticle(params);
  expect(result).toMatchSnapshot();
});

afterEach(() => {
  gql.__reset();
});
