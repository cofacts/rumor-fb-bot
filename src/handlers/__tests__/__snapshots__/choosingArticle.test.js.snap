// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should ask users if they want to submit article when user say not found 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AV8d2-YtyCdS-nWhuhdi",
    ],
    "searchedText": "這一篇文章確實是一個轉傳文章，他夠長，看起來很轉傳，但是使用者覺得資料庫裡沒有。",
    "selectedArticleId": undefined,
  },
  "event": Object {
    "input": "0",
    "message": Object {
      "id": "7045918737413",
      "text": "0",
      "type": "text",
    },
    "timestamp": 1511633232479,
    "type": "message",
  },
  "isSkipUser": false,
  "issuedAt": 1511633232970,
  "replies": Array [
    Object {
      "content": Object {
        "text": "【送出訊息到公開資料庫？】
若這是「轉傳訊息」，而且您覺得這很可能是一則「謠言」，請將這則訊息送進公開資料庫建檔，讓好心人查證與回覆。

雖然您不會立刻收到查證結果，但可以幫助到未來同樣收到這份訊息的人。",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "text": "請把「為何您會覺得這是一則謠言」的理由傳給我們，幫助闢謠編輯釐清您有疑惑之處。",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "attachment": Object {
          "payload": Object {
            "buttons": Array [
              Object {
                "payload": "n",
                "title": "放棄送出",
                "type": "postback",
              },
            ],
            "template_type": "button",
            "text": "若要放棄，請按「放棄送出」。",
          },
          "type": "template",
        },
      },
      "type": "buttons",
    },
  ],
  "state": "ASKING_ARTICLE_SUBMISSION_REASON",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;

exports[`should block user from submitting articles that is too short 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AV8d2-YtyCdS-nWhuhdi",
    ],
    "searchedText": "這真的假的？？！！！？？！",
    "selectedArticleId": undefined,
  },
  "event": Object {
    "input": "0",
    "message": Object {
      "id": "7045918737413",
      "text": "0",
      "type": "text",
    },
    "timestamp": 1511633232479,
    "type": "message",
  },
  "isSkipUser": false,
  "issuedAt": 1511633232970,
  "replies": Array [
    Object {
      "content": Object {
        "text": "剛才您傳的訊息資訊量太少，編輯無從查證。
查證範圍請參考📖使用手冊 http://bit.ly/cofacts-fb-users",
      },
      "type": "text",
    },
  ],
  "state": "__INIT__",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;

exports[`should select article and have OPINIONATED and NOT_ARTICLE replies 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AV8d2-YtyCdS-nWhuhdi",
    ],
    "foundReplyIds": Array [
      "AV--O3nfyCdS-nWhujMD",
      "AV8fXikZyCdS-nWhuhfN",
      "AV--LRZYyCdS-nWhujL9",
      "AV8jkRlByCdS-nWhuhiY",
    ],
    "searchedText": "老榮民九成存款全部捐給慈濟，如今窮了卻得不到慈濟醫院社工的幫忙，竟翻臉不認人",
    "selectedArticleId": "AV8d2-YtyCdS-nWhuhdi",
  },
  "event": Object {
    "input": "1",
    "message": Object {
      "id": "7045918737413",
      "text": "1",
      "type": "text",
    },
    "timestamp": 1511633232479,
    "type": "message",
  },
  "isSkipUser": false,
  "issuedAt": 1511633232970,
  "replies": Array [
    Object {
      "content": Object {
        "text": "這個訊息有：
1 則回應標成 ❌ 含有不實訊息
0 則回應標成 ⭕ 含有真實訊息
1 則回應標成 💬 含有個人意見
2 則回應標成 ⚠️️ 不在查證範圍
",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "payload": Object {
          "elements": Array [
            Object {
              "buttons": Array [
                Object {
                  "payload": 1,
                  "title": "閱讀此回應",
                  "type": "postback",
                },
              ],
              "subtitle": "❌ 含有不實訊息
[有 3 人覺得此回應有幫助]",
              "title": "老榮民有住所，有月退俸，也不需要特別的醫療護理，因此較適合榮民之家養老，慈濟也一直關懷陪伴。護理之家有免費方案，入住資格和有無捐款沒有關係。
慈善機構使用資源需",
            },
            Object {
              "buttons": Array [
                Object {
                  "payload": 2,
                  "title": "閱讀此回應",
                  "type": "postback",
                },
              ],
              "subtitle": "💬 含有個人意見
[還沒有人針對此回應評價]",
              "title": "慈濟自來有褒有貶，每人有各自的看法。",
            },
            Object {
              "buttons": Array [
                Object {
                  "payload": 3,
                  "title": "閱讀此回應",
                  "type": "postback",
                },
              ],
              "subtitle": "⚠️️ 不在查證範圍
[還沒有人針對此回應評價]",
              "title": "整個基督教徒的打手，自己不努力去牧人、募款，利用攻擊別人，三分真七分假，更利用多帳號來造勢及造風向，害我也差點相信了，各位有智慧的看官，看過之後，笑一笑即可，別",
            },
            Object {
              "buttons": Array [
                Object {
                  "payload": 4,
                  "title": "閱讀此回應",
                  "type": "postback",
                },
              ],
              "subtitle": "⚠️️ 不在查證範圍
[還沒有人針對此回應評價]",
              "title": "文章是假的 照片也是盜用我方家人的照片,嚴重影響我家人名譽,我方已對po文的人聯絡並打算提出告訴,請勿相信謠言.",
            },
          ],
          "template_type": "generic",
        },
        "type": "template",
      },
      "type": "carousel",
    },
  ],
  "state": "CHOOSING_REPLY",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;

exports[`should select article and slice replies when over 10 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AV8d2-YtyCdS-nWhuhdi",
    ],
    "foundReplyIds": Array [
      "AV--O3nfyCdS-nWhujMD",
      "AV--O3nfyCdS-nWhujMD",
      "AV--O3nfyCdS-nWhujMD",
      "AV--O3nfyCdS-nWhujMD",
      "AV--O3nfyCdS-nWhujMD",
      "AV--O3nfyCdS-nWhujMD",
      "AV8fXikZyCdS-nWhuhfN",
      "AV--O3nfyCdS-nWhujMD",
      "AV--O3nfyCdS-nWhujMD",
      "AV--LRZYyCdS-nWhujL9",
      "AV8jkRlByCdS-nWhuhiY",
    ],
    "searchedText": "老榮民九成存款全部捐給慈濟，如今窮了卻得不到慈濟醫院社工的幫忙，竟翻臉不認人",
    "selectedArticleId": "AV8d2-YtyCdS-nWhuhdi",
  },
  "event": Object {
    "input": "1",
    "message": Object {
      "id": "7045918737413",
      "text": "1",
      "type": "text",
    },
    "timestamp": 1511633232479,
    "type": "message",
  },
  "isSkipUser": false,
  "issuedAt": 1511633232970,
  "replies": Array [
    Object {
      "content": Object {
        "text": "這個訊息有：
6 則回應標成 ❌ 含有不實訊息
2 則回應標成 ⭕ 含有真實訊息
1 則回應標成 💬 含有個人意見
2 則回應標成 ⚠️️ 不在查證範圍
",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "payload": Object {
          "elements": Array [
            Object {
              "buttons": Array [
                Object {
                  "payload": 1,
                  "title": "閱讀此回應",
                  "type": "postback",
                },
              ],
              "subtitle": "❌ 含有不實訊息
[還沒有人針對此回應評價]",
              "title": "回應1",
            },
            Object {
              "buttons": Array [
                Object {
                  "payload": 2,
                  "title": "閱讀此回應",
                  "type": "postback",
                },
              ],
              "subtitle": "❌ 含有不實訊息
[還沒有人針對此回應評價]",
              "title": "回應2",
            },
            Object {
              "buttons": Array [
                Object {
                  "payload": 3,
                  "title": "閱讀此回應",
                  "type": "postback",
                },
              ],
              "subtitle": "⭕ 含有真實訊息
[還沒有人針對此回應評價]",
              "title": "回應3",
            },
            Object {
              "buttons": Array [
                Object {
                  "payload": 4,
                  "title": "閱讀此回應",
                  "type": "postback",
                },
              ],
              "subtitle": "⭕ 含有真實訊息
[還沒有人針對此回應評價]",
              "title": "回應4",
            },
            Object {
              "buttons": Array [
                Object {
                  "payload": 5,
                  "title": "閱讀此回應",
                  "type": "postback",
                },
              ],
              "subtitle": "❌ 含有不實訊息
[還沒有人針對此回應評價]",
              "title": "回應5",
            },
            Object {
              "buttons": Array [
                Object {
                  "payload": 6,
                  "title": "閱讀此回應",
                  "type": "postback",
                },
              ],
              "subtitle": "❌ 含有不實訊息
[還沒有人針對此回應評價]",
              "title": "回應6",
            },
            Object {
              "buttons": Array [
                Object {
                  "payload": 7,
                  "title": "閱讀此回應",
                  "type": "postback",
                },
              ],
              "subtitle": "💬 含有個人意見
[還沒有人針對此回應評價]",
              "title": "回應9",
            },
            Object {
              "buttons": Array [
                Object {
                  "payload": 8,
                  "title": "閱讀此回應",
                  "type": "postback",
                },
              ],
              "subtitle": "❌ 含有不實訊息
[還沒有人針對此回應評價]",
              "title": "回應10",
            },
            Object {
              "buttons": Array [
                Object {
                  "payload": 9,
                  "title": "閱讀此回應",
                  "type": "postback",
                },
              ],
              "subtitle": "❌ 含有不實訊息
[還沒有人針對此回應評價]",
              "title": "回應11",
            },
            Object {
              "buttons": Array [
                Object {
                  "payload": 10,
                  "title": "閱讀此回應",
                  "type": "postback",
                },
              ],
              "subtitle": "⚠️️ 不在查證範圍
[還沒有人針對此回應評價]",
              "title": "回應7。
",
            },
          ],
          "template_type": "generic",
        },
        "type": "template",
      },
      "type": "carousel",
    },
    Object {
      "content": Object {
        "text": "更多回應請到：https://cofacts.g0v.tw/article/AV8d2-YtyCdS-nWhuhdi",
      },
      "type": "text",
    },
  ],
  "state": "CHOOSING_REPLY",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;

exports[`should select article by articleId 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AVyyB61NyCdS-nWhuakC",
      "AV4sEcSHyCdS-nWhufEX",
      "5478658696099-rumor",
    ],
    "foundReplyIds": Array [
      "AVygFA0RyCdS-nWhuaXY",
      "AVy6LkWIyCdS-nWhuaqu",
    ],
    "searchedText": "《緊急通知》
台北馬偕醫院傳來訊息：
資深醫生（林清風）傳來：「請大家以後千萬不要再吃生魚片了！」
因為最近已經發現- 好多病人因為吃了生魚片，胃壁附著《海獸胃腺蟲》，大小隻不一定，有的病人甚至胃壁上滿滿都是無法夾出來，驅蟲藥也很難根治，罹患機率每個國家的人都一樣。
尤其；鮭魚的含蟲量最高、最可怕！
請傳給朋友，讓他們有所警惕!",
    "selectedArticleId": "AVyyB61NyCdS-nWhuakC",
  },
  "event": Object {
    "input": "1",
    "message": Object {
      "id": "6691804381697",
      "text": "1",
      "type": "text",
    },
    "timestamp": 1505314293406,
    "type": "message",
  },
  "isSkipUser": false,
  "issuedAt": 1505314295017,
  "replies": Array [
    Object {
      "content": Object {
        "text": "這個訊息有：
2 則回應標成 ❌ 含有不實訊息
0 則回應標成 ⭕ 含有真實訊息
0 則回應標成 💬 含有個人意見
0 則回應標成 ⚠️️ 不在查證範圍
",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "payload": Object {
          "elements": Array [
            Object {
              "buttons": Array [
                Object {
                  "payload": 1,
                  "title": "閱讀此回應",
                  "type": "postback",
                },
              ],
              "subtitle": "❌ 含有不實訊息
[有 14 人覺得此回應有幫助
有 2 人覺得此回應沒幫助]",
              "title": "查證這則謠言的過程中，約翰走鹿也發現其實網路早就有許多解答，就讓約翰走鹿來借花獻佛一下囉。多家媒體都有針對這則「吃生魚片會有海獸胃腺蟲附著胃壁」的謠言請教醫師，",
            },
            Object {
              "buttons": Array [
                Object {
                  "payload": 2,
                  "title": "閱讀此回應",
                  "type": "postback",
                },
              ],
              "subtitle": "❌ 含有不實訊息
[有 24 人覺得此回應有幫助]",
              "title": "過去台灣只有一個海獸胃腺蟲鑽入胃壁的病例，而且僅有一隻蟲。而經查詢，台北馬偕醫院也沒有名為林清風的醫師。",
            },
          ],
          "template_type": "generic",
        },
        "type": "template",
      },
      "type": "carousel",
    },
  ],
  "state": "CHOOSING_REPLY",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;

exports[`should select article with no replies 1`] = `
Object {
  "data": Object {
    "foundArticleIds": Array [
      "AV_4WX8vyCdS-nWhujyH",
    ],
    "searchedText": "老司機車裡總備一塊香皂，知道內情的新手默默也準備了一塊",
    "selectedArticleId": "AV_4WX8vyCdS-nWhujyH",
  },
  "event": Object {
    "input": "1",
    "message": Object {
      "id": "7049700770815",
      "text": "1",
      "type": "text",
    },
    "timestamp": 1511702208226,
    "type": "message",
  },
  "isSkipUser": false,
  "issuedAt": 1511702208730,
  "replies": Array [
    Object {
      "content": Object {
        "text": "【跟編輯說您的疑惑】
抱歉這篇訊息還沒有人回應過唷！

若您覺得這是一則謠言，請指出您有疑惑之處，說服編輯這是一份應該被闢謠的訊息。

",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "text": "請把「為何您會覺得這是一則謠言」的理由打字傳給我們，幫助闢謠編輯釐清您有疑惑之處；
",
      },
      "type": "text",
    },
    Object {
      "content": Object {
        "attachment": Object {
          "payload": Object {
            "buttons": Array [
              Object {
                "payload": "n",
                "title": "我不想填理由",
                "type": "postback",
              },
            ],
            "template_type": "button",
            "text": "若想跳過，請按「我不想填理由」。",
          },
          "type": "template",
        },
      },
      "type": "buttons",
    },
  ],
  "state": "ASKING_REPLY_REQUEST_REASON",
  "userId": "Uc76d8ae9ccd1ada4f06c4e1515d46466",
}
`;
